docs/git-guide.md

```diff
@@ -0,0 +1,1052 @@
+
+```
+$ git branch
+* issue-demo
+  master
+```
+
+To see even more information about your branches, including remote branches, use `git branch -vva`:
+
+```
+$ git branch -vva
+* issue-123                 517468b troubleshooting tip about provisioning
+  master                    f0eaee6 [origin/master] bug: Fix traceback in get_missed_message_token_from_address().
+  remotes/origin/HEAD       -> origin/master
+  remotes/origin/issue-1234 4aeccb7 Another test commit, with longer message.
+  remotes/origin/master     f0eaee6 bug: Fix traceback in get_missed_message_token_from_address().
+  remotes/upstream/master   dbeab6a Optimize checks of test database state by moving into Python.
+```
+
+You can also configure [Bash][gitbook-other-envs-bash] and [Zsh][gitbook-other-envs-zsh] to display the current branch in your prompt.
+
+### Keep your fork up to date
+
+You'll want to [keep your fork][github-help-sync-fork] up-to-date with changes from Zulip's main repositories.
+
+**Note about git pull**: You might be used to using `git pull` on other projects. With Zulip, because we don't use merge commits, you'll want to avoid it. Rather that using `git pull`, which by default is a shortcut for `git fetch && git merge FECH_HEAD` ([docs][gitbook-git-pull]), you should use `git fetch` and then `git rebase`.
+
+First, [fetch][gitbook-fetch] changes from Zulip's upstream repository you configured in the step above:
+
+```
+$ git fetch upstream
+```
+
+Next, checkout your `master` branch and [rebase][gitbook-git-rebase] it on top of `upstream/master`:
+
+```
+$ git checkout master
+Switched to branch 'master'
+
+$ git rebase upstream/master
+```
+
+This will rollback any changes you've made to master, update it from `upstream/master`, and then re-apply your changes. Rebasing keeps the commit history clean and readable.
+
+When you're ready, [push your changes][github-help-push] to your remote fork. Make sure you're in branch `master` and the run `git push`:
+
+```
+$ git checkout master
+$ git push origin master
+```
+
+You can keep any branch up to date using this method. If you're working on a feature branch (see next section), which we recommend, you would change the command slightly, using the name of your `feature-branch` rather than `master`:
+
+```
+$ git checkout feature-branch
+Switched to branch 'feature-branch'
+
+$ git rebase upstream/master
+
+$ git push origin feature-branch
+```
+
+### Work on a feature branch
+
+One way to keep your work oraganized is to create a branch for each issue or feature.
+
+First, make sure your master branch is up-to-date with Zulip upstream ([see how][self-keep-up-to-date]).
+
+Next, from your master branch, create a new tracking branch, providing a descriptive name for your feature branch:
+
+```
+$ git checkout master
+Switched to branch 'master'
+
+$ git checkout -b issue-1755-fail2ban
+Switched to a new branch 'issue-1755-fail2ban'
```

Comment:
Yes, though you'd still need to do a `git fetch upstream` to make sure you have most recent commits. I don't have a strong preference for either method. Sometimes when writing docs aimed at beginners, I avoid compound commands, but I don't think this one has much potential for confusion.